internal/transport/remotecommand/session.go

```diff
@@ -0,0 +1,200 @@
+	OnClose() error
+}
+
+type Session interface {
+	io.Closer
+	HandleIncoming() error
+	SendStdinStream(data []byte) error
+	SendResizeStream(width, height uint16) error
+	SendCloseStdinStream() error
+}
+
+type session struct {
+	callbacks  SessionCallbacks
+	incomingCh chan []byte
+	outgoingCh chan []byte
+	mu         sync.Mutex
+	closed     bool
+	log        logrus.FieldLogger
+}
+
+func SupportedProtocols() []string {
+	return protocols
+}
+
+func NewSession(selectedProtocol string, incomingCh, outgoingCh chan []byte, callbacks SessionCallbacks, log logrus.FieldLogger) (Session, error) {
+	if !slices.Contains(protocols, selectedProtocol) {
+		return nil, fmt.Errorf("unsupported protocol: %s", selectedProtocol)
+	}
+	return &session{
+		callbacks:  callbacks,
+		incomingCh: incomingCh,
+		outgoingCh: outgoingCh,
+		log:        log,
+	}, nil
+}
+
+func (s *session) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.closed {
+		s.closed = true
+		close(s.outgoingCh)
+	}
+	return nil
+}
+
+func (s *session) sendExitCode(code int) error {
+	if s.callbacks != nil {
+		return s.callbacks.OnExit(code)
+	}
+	return nil
+}
+
+func (s *session) emitStdoutStream(message []byte) error {
+	if s.callbacks != nil {
+		return s.callbacks.OnStdout(message)
+	}
+	return nil
+}
+
+func (s *session) emitStderrStream(message []byte) error {
+	if s.callbacks != nil {
+		return s.callbacks.OnStderr(message)
+	}
+	return nil
+}
+
+func (s *session) emitErrorStream(message []byte) error {
+	status := metav1.Status{}
+	err := json.Unmarshal(message, &status)
+	if err != nil {
+		s.log.Errorf("failed to unmarshal status message: %v", err)
+		return s.sendExitCode(-1)
+	}
+	switch status.Status {
+	case metav1.StatusSuccess:
+		return s.sendExitCode(0)
+	case metav1.StatusFailure:
+		if status.Reason == remotecommand.NonZeroExitCodeReason {
+			if status.Details == nil {
+				s.log.Errorf("error stream protocol error: details must be set")
+				return s.sendExitCode(-1)
+			}
+			for i := range status.Details.Causes {
+				c := &status.Details.Causes[i]
+				if c.Type != remotecommand.ExitCodeCauseType {
+					continue
+				}
+
+				rc, err := strconv.ParseInt(c.Message, 10, 8)
+				if err != nil {
+					s.log.Errorf("error stream protocol error: invalid exit code value %q", c.Message)
+					return s.sendExitCode(-1)
+				}
+				s.log.Errorf("command terminated with exit code %d", rc)
+				return s.sendExitCode(int(rc))
+			}
+
+			s.log.Errorf("error stream protocol error: no %s cause given", remotecommand.ExitCodeCauseType)
+			return s.sendExitCode(-1)
+		}
+	default:
+		s.log.Error("error stream protocol error: unknown error")
+		return s.sendExitCode(-1)
+	}
+
+	s.log.Error(status.Message)
+	return s.sendExitCode(-1)
+}
+
+func (s *session) emitClose() {
+	if s.callbacks != nil {
+		if err := s.callbacks.OnClose(); err != nil {
+			s.log.Errorf("emitClose: %v", err)
+		}
+	}
+}
+
+func (s *session) handlePayload(payload []byte) error {
+	switch payload[0] {
```

Comment:
can payload be empty or just have a single element?

Reply:
I will add verification